MapleServer2/Managers/Actors/FieldActor.cs

```diff
@@ -5,177 +5,174 @@
 {
-    private abstract partial class FieldActor<T> : FieldObject<T>, IFieldActor<T>, IDisposable
-    {
-        public CoordF Velocity { get; set; }
-        public short Animation { get; set; }
+    public CoordF Velocity { get; set; }
+    public short Animation { get; set; }
+
+    public virtual Stats Stats { get; set; }
+    public bool IsDead { get; set; }
+
+    public List<Status> Statuses { get; set; }
+    public SkillCast SkillCast { get; set; }
+    public bool OnCooldown { get; set; }
+    public FieldNavigator Navigator { get; set; }
+    public Agent Agent { get; set; }
+
+    public FieldActor(int objectId, T value) : base(objectId, value) { }
 
-        public virtual Stats Stats { get; set; }
-        public bool IsDead { get; set; }
+    public virtual void UpdateFixed() { }
 
-        public List<Status> Statuses { get; set; }
-        public SkillCast SkillCast { get; set; }
-        public bool OnCooldown { get; set; }
-        public FieldNavigator Navigator { get; set; }
-        public Agent Agent { get; set; }
+    public virtual void Update() { }
 
-        public FieldActor(int objectId, T value) : base(objectId, value) { }
+    public virtual void MoveBy(CoordF displacement)
+    {
+        // TODO: Break displacement into segments, then set path.
+        Velocity = displacement;
+    }
 
-        public virtual void UpdateFixed() { }
+    public virtual void MoveTo(CoordF target)
+    {
+        // TODO: Perform pathfinding, then set path.
+        Velocity = target - Coord;
+    }
 
-        public virtual void Update() { }
+    public virtual void Cast(SkillCast skillCast)
+    {
+        SkillCast = skillCast;
 
-        public virtual void MoveBy(CoordF displacement)
+        // TODO: Move this and all others combat cases like recover sp to its own class.
+        // Since the cast is always sent by the skill, we have to check buffs even when not doing damage.
+        if (skillCast.IsBuffToOwner() || skillCast.IsBuffToEntity() || skillCast.IsBuffShield() || skillCast.IsDebuffToOwner())
         {
-            // TODO: Break displacement into segments, then set path.
-            Velocity = displacement;
+            Status status = new(skillCast, ObjectId, ObjectId, 1);
+            //StatusHandler.Handle(Value.Session, status);
         }
+    }
+
+    public virtual void Heal(GameSession session, Status status, int amount)
+    {
+        session.FieldManager.BroadcastPacket(SkillDamagePacket.Heal(status, amount));
+        Stats[StatAttribute.Hp].Increase(amount);
+        session.Send(StatPacket.UpdateStats(this, StatAttribute.Hp));
+    }
 
-        public virtual void MoveTo(CoordF target)
+    public virtual void RecoverHp(int amount)
+    {
+        if (amount <= 0)
         {
-            // TODO: Perform pathfinding, then set path.
-            Velocity = target - Coord;
+            return;
         }
 
-        public virtual void Cast(SkillCast skillCast)
+        lock (Stats)
```

Comment:
Why do we need locks?

Reply:
good question, I believe is to ensure thread safety since Stats is a simple Dictionary

Reply:
aren't there concurrency-safe dictionaries in C#?

Reply:
There are, I can change it in the next PR